django/utils/text.py

```diff
@@ -314,8 +316,29 @@ def phone2numeric(phone):
     return "".join(char2number.get(c, c) for c in phone.lower())
 
 
-def compress_string(s):
-    return gzip_compress(s, compresslevel=6, mtime=0)
+def _get_random_fname(max_random_bytes):
+    return b"a" * secrets.randbelow(max_random_bytes)
+
+
+def compress_string(s, *, max_random_bytes=0):
+    compressed_data = gzip_compress(s, compresslevel=6, mtime=0)
+
+    if max_random_bytes == 0:
+        return compressed_data
+
+    header = bytearray(compressed_data[:10])
+    header[3] = gzip.FNAME
+
+    fname = _get_random_fname(max_random_bytes)
+
+    return b"".join(
+        [
+            header,
+            fname,
+            b"\x00",
+            memoryview(compressed_data)[10:],
+        ]
+    )
```

Comment:
I guess you could create the `memoryview` earlier?

Also, `header[3]` contains flags, so we should, strictly speaking, use `header[3] |= gzip.FNAME`.

```suggestion
    compressed_view = memory_view(compressed_data)
    header = bytearray(compressed_view[:10])
    header[3] |= gzip.FNAME
    fname = _get_random_fname(max_random_bytes)
    return b"".join([header, fname, b"\x00", compressed_view[10:]])
```

Reply:
Also, not sure if it's cheaper, but we could do:
```python
    return header + fname + b"\x00" + compressed_view[10:]
```

Reply:
In my original code, I made put in `assert header[3] == 0` before setting the flag. I am not sure if other flags can potentially mess up the byte order of the filename or cause other problems. We do control the call to `gzip.compress` so there should never be anything else than 0 for the flags anyways. Do you have any input on that?

I originally used + instead of "".join(...) but thought that maybe join would lead to fewer allocations since python can just the build the result directly instead of doing 3 concatenations. Have not measured it and have no idea if it is better or worse. The header, fname and null termination is tiny data and the memoryview avoids copies anyways (IIUC) so I'll change to +, it is cleaner!.

Reply:
Makes sense regarding the flags, thanks for the reasoning. Also, for the join or concat, whatever is faster 🙂

Reply:
This turned into a rabbit hole!

I did some benchmarking about join vs + and concluded they were pretty much the same.

Then I found out that gzip.compress is basically a thin wrapper around GzipFile on Python 3.10. I figured we could just use GzipFile and pass filename=X to it directly to avoid all the problems of juggling it ourselves.

But then I found out that python 3.11 changed the implementation of gzip.compress to call zlib.compress directly to  improve performance: https://github.com/python/cpython/commit/ea23e7820f02840368569db8082bd0ca4d59b62a

I did some tests to compare changing the header "ourselves" vs using GzipFile(filename=...) on Python 3.11, here are the benchmarks:

```
heal-the-breach ~/code/django $ cat bench.sh
bytes=$1

echo "gzip.compress"
python -m timeit -s "from django.utils.text import compress_string; data=b'q'*$bytes" 'compress_string(data)'

echo "GzipFile"
python -m timeit -s "from django.utils.text import compress_string_gzipfile; data=b'q'*$bytes" 'compress_string_gzipfile(data)'
heal-the-breach ~/code/django $ sh bench.sh 1000
gzip.compress
50000 loops, best of 5: 5.58 usec per loop
GzipFile
20000 loops, best of 5: 12 usec per loop
heal-the-breach ~/code/django $ sh bench.sh 10000
gzip.compress
10000 loops, best of 5: 21.7 usec per loop
GzipFile
10000 loops, best of 5: 28.2 usec per loop
heal-the-breach ~/code/django $ sh bench.sh 100000
gzip.compress
1000 loops, best of 5: 242 usec per loop
GzipFile
1000 loops, best of 5: 247 usec per loop
heal-the-breach ~/code/django $ sh bench.sh 1000000
gzip.compress
100 loops, best of 5: 2.64 msec per loop
GzipFile
100 loops, best of 5: 2.64 msec per loop
heal-the-breach ~/code/django $
```

So it is actually slightly faster, at least for small sizes, to use gzip.compress directly for the sizes that we care about (1000 bytes, 10 000 bytes and 100 000 bytes).

I have adjusted the implementation to use + and create the memoryview earlier. It feels better to use the fastest implementation here since it will be called for pretty much every request. :)